Provider/ImageProvider.php

```diff
@@ -72,13 +72,44 @@ public function getHelperProperties(MediaInterface $media, $format, $options = a
             $box = $this->resizer->getBox($media, $resizerFormat);
         }
 
-        return array_merge(array(
+        $mediaWidth = $box->getWidth();
+        $src = $this->generatePublicUrl($media, $format);
+
+        $params = array(
             'alt' => $media->getName(),
             'title' => $media->getName(),
-            'src' => $this->generatePublicUrl($media, $format),
-            'width' => $box->getWidth(),
+            'src' => $src,
+            'width' => $mediaWidth,
             'height' => $box->getHeight(),
-        ), $options);
+        );
+
+        if ($format !== 'admin') {
+            $srcSet = array();
+
+            foreach ($this->getFormats() as $providerFormat => $settings) {
+                // Check if format belongs to the current media's context
+                if (strpos($providerFormat, $media->getContext()) === 0) {
+                    $width = $this->resizer->getBox($media, $settings)->getWidth();
+
+                    if ($width <= $mediaWidth) {
+                        $srcSet[] = sprintf('%s %dw', $this->generatePublicUrl($media, $providerFormat), $width);
+                    }
+                }
+            }
+
+            // The reference format is not in the formats list
+            if ($format == 'reference') {
```

Comment:
not yet, could you provide a separate PR @core23 and fix where its needed

@rande is there a reason why we don't have a constant for this?